File: reverseLL.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
// iteratively
var reverseList = function(head) {
  let prev = null;
  // point current node to previous node
  let curr = head;
    while (curr !== null) {
        // store prev node
        let nextTemp = curr.next
        // store next node before changing reference
        curr.next = prev
        prev = curr
        curr = nextTemp
    }
    // return new head 
    return prev;
};

// recursively
//
var reverseList = function(head) {
  if (head == null || head.next == null) return head
  let p = reverseList(head.next)
  head.next.next = head
  head.next = null
  return p
};

File: top-interview-q/reverseLL.py
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        curr=head
        prev=None
        while curr:
            # initialize next, then prev
            next=curr.next
            curr.next=prev
            # update prev element to curr
            prev=curr
            # update curr as next (current.next)
            curr=next
        # when curr = None and next = None, left with prev
        return prev